In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/melbourne-housing-market/Melbourne_housing_FULL.csv
/kaggle/input/melbourne-housing-market/MELBOURNE_HOUSE_PRICES_LESS.csv


In [3]:
dataset = pd.read_csv('/kaggle/input/melbourne-housing-market/Melbourne_housing_FULL.csv')

In [4]:
dataset.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [5]:
dataset.shape

(34857, 21)

In [6]:
dataset.isna().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
YearBuilt        19306
CouncilArea          3
Lattitude         7976
Longtitude        7976
Regionname           3
Propertycount        3
dtype: int64

In [8]:
dataset.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [9]:
dataset.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [12]:
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
        'Distance', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'CouncilArea',
        'Regionname', 'Propertycount']
dataset = dataset[cols_to_use]

In [13]:
dataset.shape

(34857, 15)

In [16]:
dataset.head()

Suburb  Rooms Type      Price Method SellerG  Distance  Bedroom2  \
0  Abbotsford      2    h        NaN     SS  Jellis       2.5       2.0   
1  Abbotsford      2    h  1480000.0      S  Biggin       2.5       2.0   
2  Abbotsford      2    h  1035000.0      S  Biggin       2.5       2.0   
3  Abbotsford      3    u        NaN     VB  Rounds       2.5       3.0   
4  Abbotsford      3    h  1465000.0     SP  Biggin       2.5       3.0   

   Bathroom  Car  Landsize  BuildingArea         CouncilArea  \
0       1.0  1.0     126.0           NaN  Yarra City Council   
1       1.0  1.0     202.0           NaN  Yarra City Council   
2       1.0  0.0     156.0          79.0  Yarra City Council   
3       2.0  1.0       0.0           NaN  Yarra City Council   
4       2.0  0.0     134.0         150.0  Yarra City Council   

              Regionname  Propertycount  
0  Northern Metropolitan         4019.0  
1  Northern Metropolitan         4019.0  
2  Northern Metropolitan         4019.0  
3  Northern Metropolitan         4019.0  
4  Northern Metropolitan         4019.0

In [20]:
dataset.dtypes

Suburb            object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Distance         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
CouncilArea       object
Regionname        object
Propertycount    float64
dtype: object

Now I want to check if any of the columns have missing values.

In [21]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Distance             1
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
CouncilArea          3
Regionname           3
Propertycount        3
dtype: int64

Here now I'm going to fill the missing values...
Some of them are filled by zeros - this is done to the columns of those having very few missing values and to those features that can be zero.
Some of them are filled by the mean value(mean of all the available values in that particular column) -  this is done to those columns that have high number of missing values and also to those featrees that can't be zero i.e area etc.

Also we have to see what kind the feature is...

In [27]:
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Price             7610
Method               0
SellerG              0
Distance             0
Bedroom2             0
Bathroom             0
Car                  0
Landsize         11810
BuildingArea     21115
CouncilArea          3
Regionname           3
Propertycount        0
dtype: int64

In [29]:
dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())

In [30]:
dataset.isna().sum()

Suburb              0
Rooms               0
Type                0
Price            7610
Method              0
SellerG             0
Distance            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea        0
CouncilArea         3
Regionname          3
Propertycount       0
dtype: int64

In [31]:
dataset.dropna(inplace = True)
dataset.isna().sum()

Suburb           0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Distance         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
CouncilArea      0
Regionname       0
Propertycount    0
dtype: int64

Now I do one hot encoding for the text columns.

In [32]:
dataset = pd.get_dummies(dataset, drop_first=True)
dataset.head()

Rooms      Price  Distance  Bedroom2  Bathroom  Car  Landsize  \
1      2  1480000.0       2.5       2.0       1.0  1.0     202.0   
2      2  1035000.0       2.5       2.0       1.0  0.0     156.0   
4      3  1465000.0       2.5       3.0       2.0  0.0     134.0   
5      3   850000.0       2.5       3.0       2.0  1.0      94.0   
6      4  1600000.0       2.5       3.0       1.0  2.0     120.0   

   BuildingArea  Propertycount  Suburb_Aberfeldie  ...  \
1      160.2564         4019.0                  0  ...   
2       79.0000         4019.0                  0  ...   
4      150.0000         4019.0                  0  ...   
5      160.2564         4019.0                  0  ...   
6      142.0000         4019.0                  0  ...   

   CouncilArea_Wyndham City Council  CouncilArea_Yarra City Council  \
1                                 0                               1   
2                                 0                               1   
4                                 0                               1   
5                                 0                               1   
6                                 0                               1   

   CouncilArea_Yarra Ranges Shire Council  Regionname_Eastern Victoria  \
1                                       0                            0   
2                                       0                            0   
4                                       0                            0   
5                                       0                            0   
6                                       0                            0   

   Regionname_Northern Metropolitan  Regionname_Northern Victoria  \
1                                 1                             0   
2                                 1                             0   
4                                 1                             0   
5                                 1                             0   
6                                 1                             0   

   Regionname_South-Eastern Metropolitan  Regionname_Southern Metropolitan  \
1                                      0                                 0   
2                                      0                                 0   
4                                      0                                 0   
5                                      0                                 0   
6                                      0                                 0   

   Regionname_Western Metropolitan  Regionname_Western Victoria  
1                                0                            0  
2                                0                            0  
4                                0                            0  
5                                0                            0  
6                                0                            0  

[5 rows x 745 columns]

In [33]:
dataset.shape

(27244, 745)

Now I create X and y.
X- independent feature or the input label
y - target feature and here it is the price

In [34]:
X = dataset.drop('Price', axis=1)
y = dataset.Price

In [81]:
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X,y,test_size=0.3,random_state=2)
# 30% testing and 70% data for training

In [82]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(train_X,train_y)

In [83]:
reg.score(train_X,train_y), reg.score(test_X,test_y)

(0.6827792395792723, 0.13853683161641805)

The score came out to be very low which is approximately 14% whereas the score for training dataset is 68% which clearly tell us that the model is overfitting. For training example the scoe wae fine but for test examples the score is horrible. In order to get rid over it I ,ake use of the regularisation i.e performing lasso and ridge regressions.

In [84]:
from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha = 50, max_iter = 100, tol = 0.1)
lasso_reg.fit(train_X,train_y)

/opt/conda/lib/python3.10/site-packages/sklearn/linear_model/_coordinate_descent.py:631: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.329e+15, tolerance: 7.963e+14
  model = cd_fast.enet_coordinate_descent(


Lasso(alpha=50, max_iter=100, tol=0.1)

In [85]:
lasso_reg.score(train_X,train_y), lasso_reg.score(test_X,test_y)

(0.6767356948457683, 0.6637669697137103)

Here one can find that the accuracy of test data has been improved and is very close to that of train data. Hence overfitting has been improved.

In [86]:
from sklearn.linear_model import Ridge


ridge_reg = Ridge(alpha=50, max_iter=100, tol=0.1) 
ridge_reg.fit(train_X, train_y)

Ridge(alpha=50, max_iter=100, tol=0.1)

In [87]:
ridge_reg.score(train_X,train_y), ridge_reg.score(test_X, test_y) 

(0.6622376739684328, 0.6670848945194958)

Here we can clearly see that the accuracy of the model with test data has further increased to 66.7%

So by using Lasso and Ridge regressions, the accuracy for the not seen datasamples(test data) improved a whole lot.